internal/replication/utils.go

```diff
@@ -28,7 +30,86 @@ func raftEntryString(e raftpb.Entry) string {
 		e.Term, e.Index, e.Type, len(e.Data), h.Sum64())
 }
 
+// VerifyConsensusReConfig checks the configuration changes in types.ConsensusConfig.
+//
+// This method checks that the changes between one ConsensusConfig to the next are safe, because some mutations might
+// cause a permanent loss of quorum in the cluster, something that is very difficult to recover from.
+// - Members can be added or removed (membership change) one member at a time
+// - Members' endpoints cannot be changed together with a membership change
+// - An existing member cannot change its Raft ID (it must be removed from the cluster and added again as a new member)
+// - The Raft ID of a new member must be unique - therefore it must be larger than MaxRaftId
+//
+// We assume that both the current and updated ClusterConfig are internally consistent, specifically, that the Nodes
+// and the ConsensusConfig.Members arrays match by NodeId in each.
+func VerifyConsensusReConfig(currentConfig, updatedConfig *types.ConsensusConfig, lg *logger.SugarLogger) error {
+	addedPeers, removedPeers, changedPeers, err := detectPeerConfigChanges(currentConfig, updatedConfig)
+	if err != nil {
+		return err
+	}
+
+	if len(addedPeers)+len(removedPeers) > 1 {
+		return errors.Errorf("cannot make more than one membership change at a time: %d added, %d removed", len(addedPeers), len(removedPeers))
+	}
+
+	if (len(addedPeers)+len(removedPeers) == 1) && len(changedPeers) > 1 {
+		return errors.Errorf("cannot update peer endpoints while making membership changes: %d added, %d removed, %d updated", len(addedPeers), len(removedPeers), len(changedPeers))
+	}
+
+	if len(addedPeers)+len(removedPeers) == 1 {
+		// TODO support dynamic membership changes, see:
+		return errors.New("dynamic membership changes to the cluster are not supported yet")
+	}
+
+	if !proto.Equal(currentConfig.RaftConfig, updatedConfig.RaftConfig) {
+		lg.Warning("ConsensusConfig RaftConfig changed, the new RaftConfig will be applied after server restart!")
+	}
+
+	return nil
+}
+
+func detectPeerConfigChanges(currentConfig, updatedConfig *types.ConsensusConfig) (addedPeers, removedPeers, changedPeers []*types.PeerConfig, err error) {
+	currPeers := make(map[string]*types.PeerConfig)
+	for _, m := range currentConfig.Members {
+		currPeers[m.NodeId] = m
+	}
+
+	updtPeers := make(map[string]*types.PeerConfig)
+	for _, updtMember := range updatedConfig.Members {
+		updtPeers[updtMember.NodeId] = updtMember
+		if currMember := currPeers[updtMember.NodeId]; currMember != nil {
```

Comment:
I think most go examples use "comma ok" idiom, because it defends against "zero value". 
`if currMember, ok := currPeers[updtMember.NodeId]; ok {`
In this specific case it "zero value" should be `nil`, so it doesn't matter much, but lets make it consistent.

Reply:
Done